CASSGO-122 Panic when using a HostFilter and keyspace is not replicated to every DC

Hey, we see this panic

```
panic: token map different size to token ring: got 0 expected 120
goroutine 415 [running]:
github.com/apache/cassandra-gocql-driver/v2.(*networkTopology).replicaMap(0x25c6b9516810, 0x25c6b96629c0)
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/topology.go:391 +0x1274
github.com/apache/cassandra-gocql-driver/v2.(*tokenAwareHostPolicy).updateAllReplicas(0x25c6b8c53ac8?, 0x25c6b87a1de0, 0xd326c0?)
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/policies.go:559 +0x112
github.com/apache/cassandra-gocql-driver/v2.(*tokenAwareHostPolicy).schemaRefreshed(0x25c6b8b95800, 0x25c6b8782dc0)
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/policies.go:543 +0x90
github.com/apache/cassandra-gocql-driver/v2.refreshSchemas(0x25c6b8694f08)
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/metadata.go:865 +0xd71
github.com/apache/cassandra-gocql-driver/v2.NewSession.func1()
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/session.go:174 +0x17
github.com/apache/cassandra-gocql-driver/v2.(*refreshDebouncer).flusher(0x25c6b85b3590)
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/host_source.go:1009 +0x17d
created by github.com/apache/cassandra-gocql-driver/v2.newRefreshDebouncer in goroutine 414
	/go/pkg/mod/github.com/apache/cassandra-gocql-driver/v2@v2.1.0/host_source.go:945 +0x145
```

It happens when session is configured with host filter to connect only to hosts in a local dc:
```go
cluster.HostFilter = gocql.HostFilterFunc(func(host *gocql.HostInfo) bool {
	return host.DataCenter() == cfg.LocalDC
})
```

And  there is a keyspace that is not replicated to all DCs:
```
CREATE KEYSPACE ks1 WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '3'}  AND durable_writes = true;
CREATE KEYSPACE ks2 WITH replication = {'class': 'NetworkTopologyStrategy', 'dc1': '3', 'dc2': '3'}  AND durable_writes = true;
```

UPD. It happens for clients with LocalDC: dc2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [2.1.2]
+
+### Fixed
+
+- Prevent panic when using a HostFilter and keyspace is not replicated to every DC (CASSGO-122)
+
 ## [2.1.1]
 
 ### Fixed

--- a/policies.go
+++ b/policies.go
@@ -524,7 +524,7 @@ func (t *tokenAwareHostPolicy) updateReplicas(keyspace string) {
 				if _, ok := meta.replicas[key]; !ok {
 					metaUpdate := t.getMetadataForUpdate()
 					newReplicas := make(map[string]tokenRingReplicas, len(meta.replicas))
-					newReplicas[key] = strat.replicaMap(metaUpdate.tokenRing)
+					newReplicas[key] = strat.replicaMap(metaUpdate.tokenRing, t.logger)
 					for k, replicas := range metaUpdate.replicas {
 						newReplicas[k] = replicas
 					}
@@ -556,7 +556,7 @@ func (t *tokenAwareHostPolicy) updateAllReplicas(meta *clusterMeta, schemaMeta *
 			if meta != nil && meta.tokenRing != nil {
 				key := strat.strategyKey()
 				if _, ok := newReplicas[key]; !ok {
-					newReplicas[key] = strat.replicaMap(meta.tokenRing)
+					newReplicas[key] = strat.replicaMap(meta.tokenRing, t.logger)
 				}
 			}
 		}

--- a/topology.go
+++ b/topology.go
@@ -66,7 +66,7 @@ func (h tokenRingReplicas) replicasFor(t token) *hostTokens {
 }
 
 type placementStrategy interface {
-	replicaMap(tokenRing *tokenRing) tokenRingReplicas
+	replicaMap(tokenRing *tokenRing, logger StructuredLogger) tokenRingReplicas
 	replicationFactor(dc string) int
 	// strategyKey returns a unique identifier string for this strategy instance.
 	// Two strategy instances with identical configuration should return the same key.
@@ -162,7 +162,7 @@ func (s *simpleStrategy) replicationFactor(dc string) int {
 	return s.rf
 }
 
-func (s *simpleStrategy) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
+func (s *simpleStrategy) replicaMap(tokenRing *tokenRing, _ StructuredLogger) tokenRingReplicas {
 	tokens := tokenRing.tokens
 	ring := make(tokenRingReplicas, len(tokens))
 
@@ -255,7 +255,7 @@ func (n *networkTopology) haveRF(replicaCounts map[string]int) bool {
 	return true
 }
 
-func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
+func (n *networkTopology) replicaMap(tokenRing *tokenRing, logger StructuredLogger) tokenRingReplicas {
 	dcRacks := make(map[string]map[string]struct{}, len(n.dcs))
 	// skipped hosts in a dc
 	skipped := make(map[string][]*HostInfo, len(n.dcs))
@@ -324,7 +324,11 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 				continue
 			} else if replicasInDC[dc] >= rf {
 				if replicasInDC[dc] > rf {
-					panic(fmt.Sprintf("replica overflow. rf=%d have=%d in dc %q", rf, replicasInDC[dc], dc))
+					logger.Warning("Replica overflow. Returning empty map.",
+						NewLogFieldInt("rf", rf),
+						NewLogFieldInt("have", replicasInDC[dc]),
+						NewLogFieldString("dc", dc))
+					return tokenRingReplicas{}
 				}
 
 				// have enough replicas in this DC
@@ -372,23 +376,36 @@ func (n *networkTopology) replicaMap(tokenRing *tokenRing) tokenRingReplicas {
 		}
 
 		if len(replicas) == 0 {
-			panic(fmt.Sprintf("no replicas for token: %v", th.token))
+			logger.Warning("No replicas for token. Returning empty map.",
+				NewLogFieldString("token", th.token.String()))
+			return tokenRingReplicas{}
 		} else if !replicas[0].Equal(th.host) {
-			panic(fmt.Sprintf("first replica is not the primary replica for the token: expected %v got %v", replicas[0].ConnectAddress(), th.host.ConnectAddress()))
+			logger.Warning("First replica is not the primary replica for the token. Returning empty map.",
+				NewLogFieldString("token", th.token.String()),
+				NewLogFieldIP("expected", replicas[0].ConnectAddress()),
+				NewLogFieldIP("got", th.host.ConnectAddress()))
+			return tokenRingReplicas{}
 		}
 
 		replicaRing = append(replicaRing, hostTokens{th.token, replicas})
 	}
 
 	dcsWithReplicas := 0
-	for _, dc := range n.dcs {
-		if dc > 0 {
+	for dc, rf := range n.dcs {
+		// We should count only DCs that driver is aware of and have a replication factor > 0
+		if _, knownDc := dcRacks[dc]; knownDc && rf > 0 {
 			dcsWithReplicas++
 		}
 	}
 
 	if dcsWithReplicas == len(dcRacks) && len(replicaRing) != len(tokens) {
-		panic(fmt.Sprintf("token map different size to token ring: got %d expected %d", len(replicaRing), len(tokens)))
+		logger.Warning("Unexpected state while building replica map. Returning empty map.",
+			NewLogFieldString("strategy_key", n.strategyKey()),
+			NewLogFieldInt("dcs_with_replicas", dcsWithReplicas),
+			NewLogFieldInt("dcs_in_ring", len(dcRacks)),
+			NewLogFieldInt("token_ring_size", len(tokens)),
+			NewLogFieldInt("replica_ring_size", len(replicaRing)))
+		return tokenRingReplicas{}
 	}
 
 	return replicaRing

--- a/topology_test.go
+++ b/topology_test.go
@@ -31,6 +31,8 @@ import (
 	"fmt"
 	"sort"
 	"testing"
+
+	"github.com/stretchr/testify/require"
 )
 
 func TestPlacementStrategy_SimpleStrategy(t *testing.T) {
@@ -49,7 +51,7 @@ func TestPlacementStrategy_SimpleStrategy(t *testing.T) {
 	hosts := []*HostInfo{host0, host25, host50, host75}
 
 	strat := newSimpleStrategy(2)
-	tokenReplicas := strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens})
+	tokenReplicas := strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens}, nopLoggerSingleton)
 	if len(tokenReplicas) != len(tokens) {
 		t.Fatalf("expected replica map to have %d items but has %d", len(tokens), len(tokenReplicas))
 	}
@@ -157,7 +159,7 @@ func TestPlacementStrategy_NetworkStrategy(t *testing.T) {
 				expReplicas += rf
 			}
 
-			tokenReplicas := test.strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens})
+			tokenReplicas := test.strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens}, nopLoggerSingleton)
 			if len(tokenReplicas) != test.expectedReplicaMapSize {
 				t.Fatalf("expected replica map to have %d items but has %d", test.expectedReplicaMapSize,
 					len(tokenReplicas))
@@ -223,4 +225,34 @@ func TestPlacementStrategy_NetworkStrategy(t *testing.T) {
 			}
 		})
 	}
+}
+
+// Regression test for CASSGO-122:
+// when the token ring only contains hosts from a DC that has RF=0/unspecified for a keyspace,
+// networkTopology.replicaMap should return an empty replica map.
+func TestPlacementStrategy_NetworkStrategy_ReturnEmptyReplicaMapWhenNoReplicasInRing(t *testing.T) {
+	strat := newNetworkTopology(map[string]int{
+		"dc1": 3, // replicated only in dc1
+	})
+
+	// Hosts in ring only from dc2, so no replicas should be returned.
+	// hostId format: dc:rack:host which is used as a token in the token ring.
+	// It makes sense to use the hostId as a token in the token ring because it is unique and deterministic for test purpose.
+	hosts := []*HostInfo{
+		{hostId: "dc2:rack1:0", dataCenter: "dc2", rack: "rack1"},
+		{hostId: "dc2:rack2:1", dataCenter: "dc2", rack: "rack2"},
+		{hostId: "dc2:rack3:2", dataCenter: "dc2", rack: "rack3"},
+	}
+
+	tokens := make([]hostToken, 0, len(hosts))
+	for _, h := range hosts {
+		tokens = append(tokens, hostToken{
+			token: orderedToken(h.hostId),
+			host:  h,
+		})
+	}
+	sort.Sort(&tokenRing{tokens: tokens})
+
+	replicas := strat.replicaMap(&tokenRing{hosts: hosts, tokens: tokens}, nopLoggerSingleton)
+	require.Empty(t, replicas, "expected no replicas, got %d", len(replicas))
 }

